HPA v2 (API Changes)

DirectXMan12 opened the pull request:
**Release note**:
```release-note
Introduces an new alpha version of the Horizontal Pod Autoscaler including expanded support for specifying metrics.
```

Implements the API changes for kubernetes/features#117.

This implements #34754, which is the new design for the Horizontal Pod Autoscaler.  It includes improved support for custom metrics (and/or arbitrary metrics) as well as expanded support for resource metrics.  The new HPA object is introduces in the API group "autoscaling/v1alpha1".

Note that the improved custom metric support currently is limited to per pod metrics from Heapster -- attempting to use the new "object metrics" will simply result in an error.  This will change once #34586 is merged and implemented.

DirectXMan12 commented:
cc @derekwaynecarr

DirectXMan12 commented:
cc @kubernetes/autoscaling

thockin commented:
*[cmd/kube-controller-manager/app/controllermanager.go, line 395 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36033#-KWEfYRty_JuOGF7x1Yi:-KWEfYRty_JuOGF7x1Yj:b-im8603) ([raw file](https://github.com/kubernetes/kubernetes/blob/49e7d640d99a380f5163dd426343a23945dfca6a/cmd/kube-controller-manager/app/controllermanager.go#L395)):*
> ```Go
> 				metrics.DefaultHeapsterPort,
> 			)
> 			go podautoscaler.NewHorizontalController(hpaClient.Core(), hpaClient.Extensions(), hpaClient.Autoscaling(), metricsClient, s.HorizontalPodAutoscalerSyncPeriod.Duration).
> ```

halfhdskjfsh

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36033)*
<!-- Sent from Reviewable.io -->

DirectXMan12 committed:
Introduce HPA v2 API Objects

This commit introduces the autoscaling/v2alpha1 API group, which
currently contains the first alpha of the new HorizontalPodAutoscaler
object.

DirectXMan12 committed:
Update kubectl to work with new HPA objects

This commit updates kubectl to work with the changes to the unversioned
HPA object made to support the HPA v2alpha1 API.

smarterclayton commented:
Wut?

On Thu, Nov 10, 2016 at 2:20 PM, Tim Hockin <notifications@github.com>
wrote:

> *cmd/kube-controller-manager/app/controllermanager.go, line 395 at r4
> <https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36033#-KWEfYRty_JuOGF7x1Yi:-KWEfYRty_JuOGF7x1Yj:b-im8603>
> (raw file
> <https://github.com/kubernetes/kubernetes/blob/49e7d640d99a380f5163dd426343a23945dfca6a/cmd/kube-controller-manager/app/controllermanager.go#L395>):*
>
>               metrics.DefaultHeapsterPort,
>           )
>           go podautoscaler.NewHorizontalController(hpaClient.Core(), hpaClient.Extensions(), hpaClient.Autoscaling(), metricsClient, s.HorizontalPodAutoscalerSyncPeriod.Duration).
>
> halfhdskjfsh
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/36033>*
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/36033#issuecomment-259781739>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_pxhagb6R0n4jrdQPhDiMzy9-VZTBks5q825ogaJpZM4Km093>
> .
>

thockin commented:
Sorry this was during the reviewable demo, and I hit save by accident, I
think.

On Wed, Nov 30, 2016 at 6:38 PM, Clayton Coleman <notifications@github.com>
wrote:

> Wut?
>
> On Thu, Nov 10, 2016 at 2:20 PM, Tim Hockin <notifications@github.com>
> wrote:
>
> > *cmd/kube-controller-manager/app/controllermanager.go, line 395 at r4
> > <https://reviewable.kubernetes.io:443/reviews/
> kubernetes/kubernetes/36033#-KWEfYRty_JuOGF7x1Yi:-KWEfYRty_
> JuOGF7x1Yj:b-im8603>
> > (raw file
> > <https://github.com/kubernetes/kubernetes/blob/
> 49e7d640d99a380f5163dd426343a23945dfca6a/cmd/kube-controller-manager/app/
> controllermanager.go#L395>):*
> >
> > metrics.DefaultHeapsterPort,
> > )
> > go podautoscaler.NewHorizontalController(hpaClient.Core(),
> hpaClient.Extensions(), hpaClient.Autoscaling(), metricsClient, s.
> HorizontalPodAutoscalerSyncPeriod.Duration).
> >
> > halfhdskjfsh
> > ------------------------------
> >
> > *Comments from Reviewable
> > <https://reviewable.kubernetes.io:443/reviews/
> kubernetes/kubernetes/36033>*
> >
> > —
> > You are receiving this because you were assigned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/kubernetes/kubernetes/pull/
> 36033#issuecomment-259781739>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/ABG_
> pxhagb6R0n4jrdQPhDiMzy9-VZTBks5q825ogaJpZM4Km093>
> > .
> >
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/36033#issuecomment-264062676>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVIcnqQKWet59OEAb28-yOiZCYzf4ks5rDjMegaJpZM4Km093>
> .
>

DirectXMan12 committed:
Generated Files for autoscaling/v2alpha1

This commit contains all the autogenerated file changes from the commit
introducing the autoscaling/v2alpha1 API group.

DirectXMan12 committed:
Add autoscaling/v2alpha1 types to v1 in annotation

This commit adds autoscaling/v2alpha1 types to autoscaling/v1
for use in the alpha annotations which preserve v2alpha1 content
through round-trips.

DirectXMan12 commented:
@k8s-bot verify test this

jszczepkowski commented on pkg/apis/autoscaling/v1/conversion.go:
```diff
@@ -0,0 +1,200 @@
+/*
```
Do you have unittest for conversions? I couldn't find it.

DirectXMan12 replied:
they're tested by the fuzzer, IIRC, and by a host of other tests, but I can add one if needed.

mwielgus replied:
Yes, please - fuzzer will not test that the metrics are correctly converted (correct struct populated).

jszczepkowski commented on pkg/apis/autoscaling/annotations.go:
```diff
@@ -0,0 +1,25 @@
+
+// OtherMetricsAnnotation is the annotation which holds non-CPU-utilization HPA metric
+// specs when converting the `Metrics` field from autoscaling/v2alpha1
+const OtherMetricsAnnotation = "autoscaling.alpha.kubernetes.io/metrics"
```
The current names OtherMetricsAnnotation and CurrentMetricsAnnotation are similar and easy to confuse.
I would change the names so that it is explicit which one of them is status and which one of the is spec, e.g.: `OtherMetricsAnnotation`  -> `MetricsSpecAnnotation` and `CurrentMetricsAnnotation` -> `MetricsStatusAnnotation`.

DirectXMan12 replied:
ack, that seems reasonable.

jszczepkowski commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -127,3 +101,120 @@ func ValidateHorizontalPodAutoscalerStatusUpdate(newAutoscaler, oldAutoscaler *a
+	typesPresent := sets.NewString()
+	if spec.Object != nil {
+		typesPresent.Insert("object")
+		if typesPresent.Len() == 1 {
```
I don't understand this if. It should be always true (we inserted "object" to an empty set `typesPresent`).

DirectXMan12 replied:
if the user sets multiple fields in (e.g. both `object` and `resource`), the latter if will have `typesPresent.Len() == 2`, so we don't bother trying to validate the latter field.  It's a pattern found elsewhere for the union types, IIRC.

jszczepkowski replied:
But `typePresent` is initialized in line 132 as an empty set. Then you put one value to it in line 135. So, this condition is always true.

DirectXMan12 replied:
yeah, the very first one is always true, but the latter ones might not be.  Better to have all be consistent.

jszczepkowski commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -127,3 +101,120 @@ func ValidateHorizontalPodAutoscalerStatusUpdate(newAutoscaler, oldAutoscaler *a
+		allErrs = append(allErrs, field.Required(fldPath.Child("metricName"), "must specify a metric name"))
+	}
+
+	if src.TargetValue.IsZero() {
```
Shouldn't we check, if TargetValue is positive?

DirectXMan12 replied:
yep

jszczepkowski commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -127,3 +101,120 @@ func ValidateHorizontalPodAutoscalerStatusUpdate(newAutoscaler, oldAutoscaler *a
+		allErrs = append(allErrs, field.Required(fldPath.Child("metricName"), "must specify a metric name"))
+	}
+
+	if src.TargetAverageValue.IsZero() {
```
Shouldn't we check if TargetAverageValue is positive?

DirectXMan12 replied:
yep

jszczepkowski commented on pkg/apis/autoscaling/validation/validation.go:
```diff
@@ -127,3 +101,120 @@ func ValidateHorizontalPodAutoscalerStatusUpdate(newAutoscaler, oldAutoscaler *a
+
+	if src.TargetAverageUtilization != nil && src.TargetAverageValue != nil {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("targetAverageValue"), "may not set both a target raw value and a target utilization"))
+	}
```
Shouldn't we check, if src.TargetAverageValue is positive?

DirectXMan12 replied:
yep

jszczepkowski commented on pkg/apis/autoscaling/validation/validation_test.go:
```diff
@@ -244,80 +352,242 @@ func TestValidateHorizontalPodAutoscaler(t *testing.T) {
+					},
+				},
+			},
+			msg: "target.kind: Required",
```
nit: s/Required/required

DirectXMan12 replied:
IIRC, the validation errors use uppercase `Required`

jszczepkowski commented on pkg/apis/autoscaling/validation/validation_test.go:
```diff
@@ -244,80 +352,242 @@ func TestValidateHorizontalPodAutoscaler(t *testing.T) {
 				},
 			},
-			msg: "custom metric target value must be greater than 0",
+			msg: "target.kind: Required",
```
Don't you already have this test in line 491?

DirectXMan12 replied:
yes

jszczepkowski commented on pkg/apis/extensions/v1beta1/conversion.go:
```diff
@@ -47,6 +47,8 @@ func addConversionFuncs(scheme *runtime.Scheme) error {
 		Convert_v1beta1_HorizontalPodAutoscalerSpec_To_autoscaling_HorizontalPodAutoscalerSpec,
+		Convert_autoscaling_HorizontalPodAutoscalerStatus_To_v1beta1_HorizontalPodAutoscalerStatus,
+		Convert_v1beta1_HorizontalPodAutoscalerStatus_To_autoscaling_HorizontalPodAutoscalerStatus,
 	)
```
Don't we need functions:
Convert_autoscaling_HorizontalPodAutoscaler_To_v1beta1_HorizontalPodAutoscaler and Convert_v1beta1_HorizontalPodAutoscaler_To_autoscaling_HorizontalPodAutoscaler?

jszczepkowski commented on pkg/apis/extensions/v1beta1/conversion.go:
```diff
@@ -47,6 +47,8 @@ func addConversionFuncs(scheme *runtime.Scheme) error {
 		Convert_autoscaling_HorizontalPodAutoscalerSpec_To_v1beta1_HorizontalPodAutoscalerSpec,
 		Convert_v1beta1_HorizontalPodAutoscalerSpec_To_autoscaling_HorizontalPodAutoscalerSpec,
+		Convert_autoscaling_HorizontalPodAutoscalerStatus_To_v1beta1_HorizontalPodAutoscalerStatus,
+		Convert_v1beta1_HorizontalPodAutoscalerStatus_To_autoscaling_HorizontalPodAutoscalerStatus,
```
Don't we need functions:
Convert_autoscaling_HorizontalPodAutoscaler_To_v1beta1_HorizontalPodAutoscaler and Convert_v1beta1_HorizontalPodAutoscaler_To_autoscaling_HorizontalPodAutoscaler?

DirectXMan12 replied:
no, those are populated by autoconvert, since the HPA object itself isn't different (it's just meta, spec, status in both cases, just the spec and status differ)

jszczepkowski commented on pkg/apis/autoscaling/v1/conversion.go:
```diff
@@ -47,16 +47,27 @@ func Convert_autoscaling_HorizontalPodAutoscaler_To_v1_HorizontalPodAutoscaler(i
 	}
 
-	otherMetrics := make([]autoscaling.MetricSpec, 0, len(in.Spec.Metrics))
+	otherMetrics := make([]MetricSpec, 0, len(in.Spec.Metrics))
```
Is there any unittest for this?

jszczepkowski commented on pkg/kubectl/resource_printer.go:
```diff
@@ -2136,22 +2136,68 @@ func printDeploymentList(list *extensions.DeploymentList, w io.Writer, options P
 	return nil
 }
 
+func formatHPAMetrics(specs []autoscaling.MetricSpec, statuses []autoscaling.MetricStatus) string {
```
Can you please add a unit test for this?

DirectXMan12 replied:
yep

jszczepkowski commented on pkg/kubectl/describe.go:
```diff
@@ -2150,13 +2150,43 @@ func (d *HorizontalPodAutoscalerDescriber) Describe(namespace, name string, desc
-				w.Write(LEVEL_0, "%d%%\n", *hpa.Status.CurrentCPUUtilizationPercentage)
-			} else {
-				w.Write(LEVEL_0, "<unset>\n")
+		fmt.Fprintf(out, "Metrics:\t( current / target )\n")
```
Can you please add a unit test for it (file descibe_test.go)?

DirectXMan12 replied:
yep

jszczepkowski commented:
This PR is really huge. My advice: move the modifications in HPA (the last commit) to a separate PR? So, we will have two PRs: one with the API modifications and the other with the implementation.

jszczepkowski commented:
Maybe it will be even better to split this PR into 3 parts: API, kubectl and HPA implementation.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36033)
<!-- Reviewable:end -->

DirectXMan12 commented:
> Maybe it will be even better to split this PR into 3 parts: API, kubectl and HPA implementation.

I *could* do that (although support for the API without support from kubectl might cause issues) -- it's already broken up that way by commits.

DirectXMan12 commented:
@jszczepkowski kube-bot unassigned you for some reason, but I can't reassign you

DirectXMan12 commented:
I've updated this PR just to include the API object changes, and I'll do a follow-up PR with the HPA controller changes.

mwielgus commented on pkg/apis/autoscaling/v2alpha1/defaults.go:
```diff
@@ -0,0 +1,48 @@
+	}
+
+	if len(obj.Spec.Metrics) == 0 {
+		utilizationDefaultVal := int32(80)
```
Please make it a constant.

mwielgus commented on pkg/apis/autoscaling/types.go:
```diff
@@ -100,10 +182,65 @@ type HorizontalPodAutoscalerStatus struct {
-	CurrentCPUUtilizationPercentage *int32
+	CurrentAverageUtilization *int32
+	// the current value of the resource metric as a raw value
+	CurrentAverageValue resource.Quantity
```
Is it always populated?

DirectXMan12 replied:
yes, as per the proposal

mwielgus commented on pkg/apis/autoscaling/v1/conversion.go:
```diff
@@ -0,0 +1,200 @@
+	}
+
+	if len(otherMetrics) > 0 {
+		otherMetricsEnc, err := json.Marshal(otherMetrics)
```
Is there a reason why we are not using schema/api marshalers and unmarshalers there?

DirectXMan12 replied:
`pkg/api/v1/conversion.go` just uses `json.Marshal`, so I figured that was the way to go.

mwielgus commented on pkg/controller/podautoscaler/replica_calculator.go:
```diff
@@ -133,7 +134,100 @@ func (c *ReplicaCalculator) GetResourceReplicas(currentReplicas int32, targetUti
+	readyPodCount := 0
+	unreadyPods := sets.NewString()
+	missingPods := sets.NewString()
+
```
Are we implementing some new strategy here?

DirectXMan12 replied:
we need a separate function for when we're calculating resource-based replicas, but not doing percentage-of-request calculations

DirectXMan12 commented:
integration flaked on #40772

jszczepkowski commented:
LGTM from me

smarterclayton commented on pkg/apis/autoscaling/types.go:
```diff
@@ -77,10 +79,90 @@ type HorizontalPodAutoscalerSpec struct {
+
+var (
+	// a metric describing a kubernetes object (for example, hits-per-second on an Ingress object)
+	ObjectSourceType MetricSourceType = "Object"
```
Name these `ObjectMetricsSource`, `PodMetricSource` to be consistent with other constants.

DirectXMan12 replied:
`ObjectMetricSource` is the name of the actual struct, so the names would collide...

smarterclayton replied:
`ObjectMetricSourceReference`

smarterclayton replied:
Actually, I see that it has a reference in it.  Hrm.  Then use `ObjectMetricSourceType`

smarterclayton commented on pkg/apis/autoscaling/v2alpha1/types.go:
```diff
@@ -0,0 +1,237 @@
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// behaviour of autoscaler. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status.
```
The godoc for public types must follow the standard conventions - complete sentences, use the lower case field name to describe the type.  Look at other examples for framing.

DirectXMan12 replied:
yeah, that godoc was copied from the `v1`.  I'll fix it.

smarterclayton commented on pkg/apis/autoscaling/v2alpha1/types.go:
```diff
@@ -0,0 +1,237 @@
+// +genclient=true
+
+// configuration of a horizontal pod autoscaler.
+type HorizontalPodAutoscaler struct {
```
Needs to be a lot better godoc.

DirectXMan12 replied:
heh, yeah, agreeded.  It's copied from the `v1` godoc ;-)

smarterclayton reviewed (changes requested):
Overall looks pretty good, please fix the cleanup and if there's no objection i'll mark this API as approved based on the previous review from the sig.

DirectXMan12 commented:
@smarterclayton godoc look better?

DirectXMan12 commented:
@k8s-bot gce etcd3 e2e test this

smarterclayton commented:
With SIG lgtm, api review

/approve
/lgtm

sttts commented on pkg/apis/autoscaling/v1/conversion.go:
```diff
@@ -0,0 +1,239 @@
+
+	// autoscaling/v1 formerly had an implicit default applied in the controller.  In v2alpha1, we apply it explicitly.
+	// We apply it here, explicitly, since we have access to the full set of metrics from the annotation.
+	if len(out.Spec.Metrics) == 0 {
```
don't get the argument: why is this not in defaulting anymore?

sttts replied:
If I get this correctly, we apply the default depending on annotations (which store the whole `Metrics` slice in v1). This is a pretty strange logic, looking like magic behavior if you only see the v1 version as a projection of v2alpha1.

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -2145,22 +2145,75 @@ func printDeploymentList(list *extensions.DeploymentList, w io.Writer, options P
+				if len(statuses) > i && statuses[i].Resource != nil {
+					current = statuses[i].Resource.CurrentAverageValue.String()
+				}
+				list = append(list, fmt.Sprintf("%s / %s", current, spec.Resource.TargetAverageValue.String()))
```
@DirectXMan12 @jszczepkowski This output violates kubectl output conventions.

https://github.com/kubernetes/community/blob/master/contributors/devel/kubectl-conventions.md#output-conventions

The spaces around the `/` make it inconvenient for simple processing.

shiywang replied:
@pwittrock could you please take a look of pr: https://github.com/kubernetes/kubernetes/pull/56331? submitted a fix there.

bgrant0607 commented on pkg/kubectl/resource_printer.go:
```diff
@@ -2145,22 +2145,75 @@ func printDeploymentList(list *extensions.DeploymentList, w io.Writer, options P
+				if len(statuses) > i && statuses[i].Resource != nil {
+					current = statuses[i].Resource.CurrentAverageValue.String()
+				}
+				list = append(list, fmt.Sprintf("%s / %s", current, spec.Resource.TargetAverageValue.String()))
```
cc @pwittrock
